qutebrowser/browser/webkit/network/pac.py

```diff
@@ -0,0 +1,625 @@
+
+        Args:
+            query: QNetworkProxyQuery.
+
+        Return:
+            A list of QNetworkProxy objects in order of preference.
+        """
+        result = self._resolver.call([query.url().toString(),
+                                      query.peerHostName()])
+        result_str = result.toString()
+        if not result.isString():
+            err = "Got strange value from FindProxyForURL: '{}'"
+            raise EvalProxyError(err.format(result_str))
+        return self._parse_proxy_string(result_str)
+
+
+class PACFetcher(QObject):
+
+    """Asynchronous fetcher of PAC files."""
+
+    def __init__(self, url, parent=None):
+        """Resolve a PAC proxy from URL.
+
+        Args:
+            url: QUrl of a PAC proxy.
+        """
+        pac_prefix = "pac+"
+
+        super().__init__(parent)
+
+        assert url.scheme().startswith(pac_prefix)
+        url.setScheme(url.scheme()[len(pac_prefix):])
+
+        self._manager = QNetworkAccessManager()
+        self._manager.setProxy(QNetworkProxy(QNetworkProxy.NoProxy))
+        self._reply = self._manager.get(QNetworkRequest(url))
+        self._reply.finished.connect(self._finish)
+        self._pac = None
+        self._loop = None
+
+    @pyqtSlot()
+    def _finish(self):
+        if self._reply.error() != QNetworkReply.NoError:
+            log.network.error("Can't fetch PAC file from URL")
+        else:
+            try:
+                pacscript = bytes(self._reply.readAll()).decode()
+            except UnicodeError:
+                log.network.exception("Invalid encoding of a PAC file")
+            try:
+                self._pac = PACResolver(pacscript)
+                log.network.debug("Successfully evaluated PAC file.")
+            except EvalProxyError as e:
+                error = "Error in PAC evaluation: {!s}.".format(e)
+                log.network.exception(error)
+        if self._loop is not None:
+            self._reply.deleteLater()
+        else:
+            self._manager = None
+            del self._reply
+
+    def _wait(self):
+        if self._manager is not None:
+            self._loop = QEventLoop()
+            self._reply.destroyed.connect(self._loop.quit)
+            self._loop.exec()
+            self._loop = None
+            self._manager = None
+            del self._reply
+
+    def is_fetched(self):
+        """Check if PAC script is successfully fetched.
+
+        Return: True iff PAC script is downloaded and evaluated successfully.
+        """
+        self._wait()
+        return self._pac is not None
+
+    def resolve(self, query):
+        """Resolve a query via PAC.
+
+        Args: QNetworkProxyQuery.
+
+        Return: A list of QNetworkProxy objects in order of preference.
+        """
+        self._wait()
+        try:
+            return self._pac.resolve(query)
+        except (EvalProxyError, ParseProxyError) as e:
+            log.network.exception("Error in PAC resolution: {!s}.".format(e))
```

Comment:
I don't think the `!s` is needed, `str.format` already calls `str()` on the arguments.